app/models/host/managed.rb

```diff
@@ -231,6 +239,12 @@ def self.model_name
     ActiveModel::Name.new(Host)
   end
 
+  def self.smart_proxy_ids(hosts_scope = Host::Managed.where(nil))
+    hosts_scope.eager_load(smart_proxy_references).map do |host|
+      smart_proxy_references.map { |ref| host.send(ref).try(:id) }.flatten.uniq.compact
```

Comment:
`.flat_map` would be much more efficient then `.map{...}.flatten` if there is only one level of nesting.